app/src/main/java/de/tum/in/tumcampusapp/component/tumui/grades/GradesActivity.kt

```diff
@@ -0,0 +1,394 @@
+            pieMenuItem?.isVisible = true
+
+            barChartView.visibility = View.VISIBLE
+            pieChartView.visibility = View.GONE
+        }
+
+        isFetched = false
+        loadGrades(CacheControl.USE_CACHE)
+    }
+
+    override fun onRefresh() {
+        loadGrades(CacheControl.BYPASS_CACHE)
+    }
+
+    private fun initPieChart(gradeDistribution: ArrayMap<String, Int>) {
+        val entries = grades.map { grade ->
+            val count = gradeDistribution[grade] ?: 0
+            PieEntry(count.toFloat(), grade)
+        }
+
+        val set = PieDataSet(entries, getString(R.string.grades_without_weight)).apply {
+            setColors(GRADE_COLORS, this@GradesActivity)
+            setDrawValues(false)
+        }
+
+        pieChartView.apply {
+            data = PieData(set)
+            setDrawEntryLabels(false)
+            legend.isWordWrapEnabled = true
+            description = null
+            invalidate()
+        }
+    }
+
+    private fun initBarChart(gradeDistribution: ArrayMap<String, Int>) {
+        val entries = grades.mapIndexed { index, grade ->
+            val value = gradeDistribution[grade] ?: 0
+            BarEntry(index.toFloat(), value.toFloat())
+        }
+
+        val set = BarDataSet(entries, getString(R.string.grades_without_weight)).apply {
+            setColors(GRADE_COLORS, this@GradesActivity)
+        }
+
+        barChartView.apply {
+            data = BarData(set)
+            setFitBars(true)
+
+            xAxis.apply {
+                granularity = 1f
+                setValueFormatter { value, _ -> grades[value.toInt()] }
+            }
+
+            description = null
+            invalidate()
+        }
+    }
+
+    /**
+     * Calculates the average grade of the given exams
+     *
+     * @param exams List of exams
+     * @return Average grade
+     */
+    private fun calculateAverageGrade(exams: List<Exam>): Double {
+        val numberFormat = NumberFormat.getInstance(Locale.GERMAN)
+        val grades = exams
+                .filter { it.isPassed }
+                .map { numberFormat.parse(it.grade).toDouble() }
+
+        val gradeSum = grades.sum()
+        return gradeSum / grades.size.toDouble()
+    }
+
+    /**
+     * Calculates grade distribution
+     *
+     * @param exams List of exams
+     * @return HashMap with grade to grade count mapping
+     */
+    private fun calculateGradeDistribution(exams: List<Exam>): ArrayMap<String, Int> {
+        val gradeDistribution = ArrayMap<String, Int>()
+        exams.forEach { exam ->
+            val count = gradeDistribution[exam.grade] ?: 0
+            gradeDistribution[exam.grade] = count + 1
+        }
+        return gradeDistribution
+    }
+
+    /**
+     * Initialize the spinner for choosing between the study programs.
+     */
+    private fun initSpinner(exams: List<Exam>) {
+        // Set Spinner data
+        val filters = arrayListOf<String>(getString(R.string.all_programs))
```

Comment:
I'd use `mutableListOf()`

Reply:
Yep, that better describes the intentions 👍